ecs-cli/modules/clients/aws/cloudformation/cluster_template.go

```diff
@@ -13,8 +13,65 @@
 
 package cloudformation
 
-func GetClusterTemplate() string {
-	return cluster_template
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ecs"
+)
+
+func GetClusterTemplate(tags []*ecs.Tag, stackName string) (string, error) {
+	tagJSON, err := json.Marshal(tags)
+	if err != nil {
+		return "", err
+	}
+
+	asgTags := getASGTags(tags, stackName)
+	asgTagJSON, err := json.Marshal(asgTags)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf(cluster_template, string(tagJSON), string(asgTagJSON)), nil
+}
+
+// Autoscaling CFN tags have an additional field that determines if they are
+// propagated to the EC2 instances launched
+// ECS CLI also adds a 'Name' tag
+// (unless customer specifies a Name; only one name is allowed by the API)
+func getASGTags(tags []*ecs.Tag, stackName string) []autoscalingTag {
+	asgTags := []autoscalingTag{}
+	addName := true
+	for _, tag := range tags {
+		asgTag := autoscalingTag{
+			Key:               aws.StringValue(tag.Key),
+			Value:             aws.StringValue(tag.Value),
+			PropagateAtLaunch: true,
+		}
+		if asgTag.Key == "Name" {
+			addName = false
+		}
+		asgTags = append(asgTags, asgTag)
+	}
+
+	if addName {
+		asgTags = append(asgTags, autoscalingTag{
+			Key:               "Name",
+			Value:             fmt.Sprintf("ECS Instance - %s", stackName),
+			PropagateAtLaunch: true,
+		})
+	}
+
+	return asgTags
+}
+
+// custom struct needed because sdk's autoscaling.Tag contains additional
+// fields that aren't valid in CFN
```

Comment:
do you know if this is bc of some constraint, or if CFN is planning on supporting those additional fields in the future?

Reply:
The SDK's struct contains fields which I think aren't relevant in CFN (resource ID and type): https://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/#Tag

Also, AFAIK the SDK doesn't actually attempt to implement any types which can be marshaled into valid CloudFormation ([goformation](https://github.com/awslabs/goformation) does tho!). It just so happened that this was the case with ECS Tags, and I wanted to add a comment explaining why using an SDK struct wasn't possible for this as well.